### DISCUSSION TITLE
perf.NewReader just hangs if perCPUBuffer is too small

### DISCUSSION THREAD

**Discussion by psanford** [cilium/ebpf#561] [Q&A] [answered]
**Describe the bug**

Using the uretprobe example as my starting point, I tried to get file paths from a tracepoint using `bpf_perf_event_output`. My userspace go program just hung on `perf.Reader.Read()` call, because I had set the `perCPUBuffer` to `os.Getpagesize()` (which for me was 4096) and my events were larger than that.

**To Reproduce**

Send a payload larger than `os.Getpagesize()` via `bpf_perf_event_output`. E.g.:
```
#define PATH_MAX 4096
struct event_t {
	u32 pid;
	char str[PATH_MAX];
};
```

**Expected behavior**
Ideally `perf.Reader.Read()` would detect that the buffer is too small and error. I've not looked at how feasible that is. Alternatively, changing the example code to use the size of `Event` for `perCPUBuffer` instead of `os.Getpagesize()` would have helped me:

```
rd, err := perf.NewReader(objs.Events, int(unsafe.Sizeof(Event{})))
```

**Comment by lmb** [comment, marked as answer]
A perf ring has no "inherent sample size", you can submit samples of variable length to it if you want. Therefore the perf reader can't know whether 4096 is enough for your needs or not.

I think if you check the return code from `bpf_perf_event_output` you'll see that it returns an error because the sample was too big. You may also be able to check https://pkg.go.dev/github.com/cilium/ebpf/perf#Record LostSamples but I haven't validated that.

**Reply by psanford** [reply]
Ah, I hadn't considered different record sizes. I do still think updating the example to set the buffer size based on the event size would be helpful newcomers.

**Reply by psanford** [reply]
I couldn't figure out a way to check LostSamples since I was never getting any records from `perf.Reader.Read()` calls.

**Reply by lmb** [reply]
> I do still think updating the example to set the buffer size based on the event size would be helpful newcomers.

You want the buffer size to be >> the event size, otherwise you will experience lost samples due to the user space not being able to keep up with the BPF

**Reply by asdfsx** [reply]
@lmb I met similar problem, I defined payload & array like
```
#define DATA_MAX 1024
struct fs_event_t
{
    struct process_ctx_t process_data;
    int flags;
    ...
    u8 data[DATA_MAX];
};

struct bpf_map_def SEC("maps/fs_events") perf_events = {
    .type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
    .key_size = 0,
    .value_size = 0,
    .max_entries = 10000,
};

struct {
	__uint(type, BPF_MAP_TYPE_RINGBUF);
	__uint(max_entries, 1 << 24);
} ringbuf_events SEC(".maps");
```
And I use perf.Reader to receive the event message
```
rd, err := perf.NewReader(objs.FsEvents, os.Getpagesize())
```
when I send the event via `bpf_perf_event_output` to `perf_events`, I got the return code `-28`.
when I use `#define DATA_MAX 400`, the return code is `0`. 
I wanna know is there any limitation for the msg size? Is there any suggestion for define the message?

BTW, when I send the event via `bpf_ringbuf_reserve`,`bpf_ringbuf_submit` to `ringbuf_events` the problem won't happen.

**Reply by lmb** [reply]
What if you set both max_entries to the same value?

**Reply by asdfsx** [reply]
@lmb you mean set the BPF_MAP_TYPE_PERF_EVENT_ARRAY‘s max_entries to `1 << 24`
I will try that

**Reply by asdfsx** [reply]
@lmb 
I update the EVENT_ARRAY 
```
struct bpf_map_def SEC("maps/fs_events") fs_events = {
    .type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
    .key_size = 0,
    .value_size = 0,
    .max_entries = 1 << 24,
};
```
But nothing changed, I still got the error code `-28`

I have another problem: when I update the`fs_events` like below
```
struct bpf_map_def SEC("maps/fs_events") fs_events = {
    .type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
    .key_size = 0,
    .value_size = sizeof(struct fs_event_t),
    .max_entries = 0,
};
``` 
I got the problem `ValueSize must be zero or four for perf event array`.
I find the code in `map.go` https://github.com/cilium/ebpf/blob/8079b37aa138b38707461cb8d548edf3359d6bc2/map.go#L124-L131
Why the value & key size should be zero or four???
When should I set the value size to 0? When should I set the value size to 4?

**Reply by florianl** [reply]
The error code `-28` translates to `ENOSPC` - `No space left on device`. It sounds like, `sizeof(struct fs_event_t)` is too large and should be split or reduced. Can you provide a full working example, that reproduces your issue? From the code snippets it is hard to make assumptions.

**Reply by brycekahle** [reply]
`BPF_MAP_TYPE_PERF_EVENT_ARRAY` is special in that the actual map values contain file descriptors to mmap-ed regions. `cilium/ebpf` and other loaders take care of this detail for you. There is no need to specify the type of the object you are going to use in the perf buffer, since it can actually be variable in size. You need to handle unmarshaling when you call `Read` on the reader. If you are wanting to control the total size of the perf buffer, you use `max_entries / MaxEntries` for that.

The key size must be 4 bytes because this is an array, thus the key is an index. For both key size and value size, `0` will cause the loader to automatically set it to 4 bytes.

**Reply by brycekahle** [reply]
If you getting `-28` or `ENOSPC` when using max entries of `1 << 24`, then that means that value is too large. Please remember that a perf buffer will create a buffer *per CPU*, so the total memory usage is really approximately `(1 << 24) * numCPUs`.

**Reply by asdfsx** [reply]
Thx a lot @lmb @florianl @brycekahle
I create repo [ebpf-test](https://github.com/asdfsx/ebpf-test) and upload my test code.
I tried to capture `vfs_write` event, and get the content.
The ringbuf version seems ok, but the perfevent version is stucked.
Is there any suggestion for improving it？

**Reply by florianl** [reply]
Just hat a quick look. For perfevents you are [allocating](https://github.com/asdfsx/ebpf-test/blob/0ef4e0bf0cde04fc304840eb9be6a3e6ac002237/perfevent/main.go#L48) `os.Getpagesize()` (`perCPUBuffer`) - on most systems `os.Getpagesize()` comes down to 4096. On the same time, you try to write an event with [an element](https://github.com/asdfsx/ebpf-test/blob/0ef4e0bf0cde04fc304840eb9be6a3e6ac002237/perfevent/counter.c#L14C11-L14C14) of `u8 data[10240]`.  As 10240 >> 4096, it will not fit into the memory that is allocated for the perfevent. 

Please try to shrink or split your element, that you will write into perfevent, so it can fit `os.perCPUBuffer`. Increasing `perCPUBuffer` for `perf.NewReader()` is also possible. But as @brycekahle, please keep in mind that `perCPUBuffer` is allocted per CPU.
